src/tree/RefreshManager.ts

```diff
@@ -0,0 +1,246 @@
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AzExtTreeItem, callWithTelemetryAndErrorHandling, IActionContext, parseError, registerCommand } from '@microsoft/vscode-azext-utils';
+import * as os from 'os';
+import * as vscode from 'vscode';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import { isCancellationError } from '../runtimes/docker';
+import { AllTreePrefixes, TreePrefix } from './TreePrefix';
+
+const pollingIntervalMs = 60 * 1000; // One minute
+const eventListenerTries = 3; // The event listener will try at most 3 times to connect for events
+
+type RefreshTarget = AzExtTreeItem | TreePrefix;
+type RefreshReason = 'interval' | 'event' | 'config' | 'manual' | 'contextChange';
+
+export class RefreshManager extends vscode.Disposable {
+    private readonly disposables: vscode.Disposable[] = [];
+    private readonly cts: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+
+    public constructor() {
+        super(() => vscode.Disposable.from(...this.disposables).dispose());
+
+        // VSCode does *not* cancel by default on disposal of a CancellationTokenSource, so we need to manually cancel
+        this.disposables.unshift(new vscode.Disposable(() => this.cts.cancel()));
+
+        this.setupRefreshOnInterval();
+        this.setupRefreshOnRuntimeEvent();
+        this.setupRefreshOnConfigurationChange();
+        this.setupRefreshOnCommand();
+        this.setupRefreshOnDockerConfigurationChange();
+        this.setupRefreshOnContextChange();
+    }
+
+    private setupRefreshOnInterval(): void {
+        const timer = setInterval(async () => {
+            for (const view of AllTreePrefixes) {
+                // Skip the registries view, which does not need to be refreshed on an interval
+                if (view === 'registries') {
+                    continue;
+                }
+                await this.refresh(view, 'interval');
+            }
+        }, pollingIntervalMs);
+
+        this.disposables.push(new vscode.Disposable(
+            () => clearInterval(timer)
+        ));
+    }
+
+    private setupRefreshOnRuntimeEvent(): void {
+        void callWithTelemetryAndErrorHandling('vscode-docker.tree.eventRefresh', async (context: IActionContext) => {
+            context.errorHandling.suppressDisplay = true;
+            context.telemetry.suppressIfSuccessful = true;
+
+            // Try at most `eventListenerTries` times to (re)connect to the event stream
+            for (let i = 0; i < eventListenerTries; i++) {
+                try {
+                    const eventGenerator = await ext.streamWithDefaultShell(client =>
+                        client.getEventStream({
+                            types: ['container', 'image', 'network', 'volume'],

```

Comment:
An issue I've found is that events are getting fired for `docker exec` calls in containers, which we do a lot of for files...which forces it into a refresh loop. These exec calls need to be ignored.